swift/apple/FirezoneNetworkExtension/PacketTunnelProvider.swift

```diff
@@ -0,0 +1,73 @@
+//
+//  PacketTunnelProvider.swift
+//  (c) 2023 Firezone, Inc.
+//  LICENSE: Apache-2.0
+//
+
+import connlib
+import Dependencies
+import NetworkExtension
+import os
+
+enum PacketTunnelProviderError: String, Error {
+  case savedProtocolConfigurationIsInvalid
+  case couldNotSetNetworkSettings
+}
+
+class PacketTunnelProvider: NEPacketTunnelProvider {
+  static let logger = Logger(subsystem: "dev.firezone.firezone", category: "packet-tunnel")
+
+  private lazy var adapter = Adapter(with: self)
+
+  override func startTunnel(
+    options _: [String: NSObject]? = nil,
+    completionHandler: @escaping (Error?) -> Void
+  ) {
+    guard let tunnelProviderProtocol = self.protocolConfiguration as? NETunnelProviderProtocol else {
+      completionHandler(PacketTunnelProviderError.savedProtocolConfigurationIsInvalid)
+      return
+    }
+
+    let providerConfiguration = tunnelProviderProtocol.providerConfiguration
+    guard let portalURL = providerConfiguration?["portalURL"] as? String else {
+      completionHandler(PacketTunnelProviderError.savedProtocolConfigurationIsInvalid)
+      return
+    }
+    guard let token = providerConfiguration?["token"] as? String else {
+      completionHandler(PacketTunnelProviderError.savedProtocolConfigurationIsInvalid)
+      return
+    }
+
+    Self.logger.log("portalURL = \(portalURL, privacy: .public)")
+    Self.logger.log("token = \(token, privacy: .public)")
+
+    do {
+      // Once connlib is updated to take in portalURL and token, this call
+      // should become adapter.start(portalURL: portalURL, token: token)
+      try adapter.start { error in
+        if let error {
+          Self.logger.error("Error in adapter.start: \(error)")
+        }
+        completionHandler(error)
+      }
+    } catch {
+      completionHandler(error)
+    }
+  }
+
+  override func stopTunnel(with _: NEProviderStopReason, completionHandler: @escaping () -> Void) {
+    adapter.stop { error in
+      if let error {
+        Self.logger.error("Error in adapter.stop: \(error)")
+      }
+      completionHandler()
+    }
+
+    #if os(macOS)
+      // HACK: This is a filthy hack to work around Apple bug 32073323 (dup'd by us as 47526107).
+      // Remove it when they finally fix this upstream and the fix has been rolled out to
+      // sufficient quantities of users.
```

Comment:
Slight improvement over this hack from https://developer.apple.com/forums/thread/84920?page=2

```suggestion
      // Hacky workaround for Apple bug 32073323.
      // See discussion at https://developer.apple.com/forums/thread/84920
      Darwin.usleep(500000)
```

Reply:
This is a very interesting hack, and could be related to the issue I see in the app described in [PR#77 in firezone-apple](https://github.com/firezone/firezone-apple/pull/77#issuecomment-1611006530).

However, I don't want to include this change in this PR. I'll test it out and, if applicable, include it in a later PR.